service/worker/deployment/deployment_workflow.go

```diff
@@ -191,7 +151,70 @@ func (d *DeploymentWorkflowRunner) run() error {
 
 	d.logger.Debug("Deployment doing continue-as-new")
 	return workflow.NewContinueAsNewError(d.ctx, DeploymentWorkflow, d.DeploymentWorkflowArgs)
+}
+
+func (d *DeploymentWorkflowRunner) validateRegisterWorker(args *deploymentspb.RegisterWorkerInDeploymentArgs) error {
+	if _, ok := d.DeploymentLocalState.TaskQueueFamilies[args.TaskQueueName].GetTaskQueues()[int32(args.TaskQueueType)]; ok {
+		return errAlreadyExists
+	}
+	return nil
+}
+
+func (d *DeploymentWorkflowRunner) handleRegisterWorker(ctx workflow.Context, args *deploymentspb.RegisterWorkerInDeploymentArgs) error {
+	// Note: use ctx in here (provided by update) instead of d.ctx
+
+	// use lock to enforce only one update at a time
+	err := d.lock.Lock(ctx)
+	if err != nil {
+		d.logger.Error("Could not acquire deployment workflow lock")
+		return err
+	}
+	d.pendingUpdates++
+	defer func() {
+		d.pendingUpdates--
+		d.lock.Unlock()
+	}()
+
+	// wait until series workflow started
+	err = workflow.Await(ctx, func() bool { return d.DeploymentLocalState.StartedSeriesWorkflow })
+	if err != nil {
+		d.logger.Error("Update canceled before series workflow started")
```

Comment:
So if the series fails to create because say it exceeds a limit on the number of series, wf will return on line 126 and in here we'd return this error to matching? Not for now, but would it be possible to propagate the "series count limit exceeded" error to matching and eventually pollers somehow?

Reply:
I don't know how we're implementing those limits.. StartWorkflowExecution or SignalWithStart couldn't fail for some application-defined limit, we'd have to change it to update-with-start? In any case, yeah I'd expect this to change when we add limits

Reply:
@ShahabT  - i might be wrong but the client kickstarts an update-with-start which shall, thanks to David's changes, kickstart a series workflow. If we enforce a limit check and return an error in the series workflow definition (per namespace limit), we reach line 126 from which the whole multi-operation should fail - in that case, the update should not even be executed right? moreover, that error shall be propagated to pollers too

Reply:
I guess my question is if matching would receive a specific error message such as "series count limit exceeded" instead of a generic message such as "Update canceled before series workflow started" or any other generic error may be sent by server because of the multi-operation failed.

Reply:
There is no multi operation at this level... matching -> deployment wf is an update-with-start, but deployment wf -> deployment series wf is currently just a start. if we want to return an error, that has to change to update-with-start also.

Reply:
Also, if this deployment wf exceeds some limit, what happens to the wf? Does it stick around to tell matching "too many"? If not, then matching will try to re-create it every time, which is bad. But if it does stick around, that's using resources that shouldn't be used. Seems like we need a cache somewhere?

Reply:
I was thinking matching would remember in memory and not retry registration, at least until a few minutes.